app/src/lib/stores/repository-state-cache.ts

```diff
@@ -46,7 +46,24 @@ export class RepositoryStateCache {
   ) {
     const currentState = this.get(repository)
     const newValues = fn(currentState)
-    this.repositoryState.set(repository.hash, merge(currentState, newValues))
+    const newState = merge(currentState, newValues)
+
+    const isSameLastLocalCommit =
+      currentState.localCommitSHAs.length > 0 &&
+      newState.localCommitSHAs.length > 0 &&
+      currentState.localCommitSHAs[0] === newState.localCommitSHAs[0]
+
+    // Only keep the "is amending" state if the last local commit hasn't changed
+    // and there is no "fixing conflicts" state.
+    const newIsAmending =
+      newState.isAmending &&
+      isSameLastLocalCommit &&
+      newState.changesState.conflictState === null
+
+    this.repositoryState.set(repository.hash, {
+      ...newState,
+      isAmending: newIsAmending,
+    })
```

Comment:
If you check [my previous approach](https://github.com/desktop/desktop/pull/12353/commits/b902f8e146ada6182d7bf6e6c07034500f75a5ee), I intended to stop amending manually whenever the user started any operation that could lead to a state where the user would amend the wrong commit (for example, you switch to a completely different branch… the amend state should be cleared… and same for pulling, pushing, merging, cherry-picking, squashing…).

After thinking a bit about it, I decided to go with this other centralized and (more) future-proof approach, where the amending state will be cleared whenever:
1. The last local commit sha changes
2. The app is in a "fix conflicts" state

This should cover all our present needs, at least!

Reply:
Nice!